cmd/cli/dashboard.go

```diff
@@ -108,67 +106,43 @@ func (d *dashboardCmd) run() error {
 	// Select pod/s given the service data available
 	set := labels.Set(svc.Spec.Selector)
 	listOptions := metav1.ListOptions{LabelSelector: set.AsSelector().String()}
-	pods, err := v1ClientSet.Pods(settings.Namespace()).
-		List(context.TODO(), listOptions)
+	pods, _ := v1ClientSet.Pods(settings.Namespace()).List(context.TODO(), listOptions)
 
 	// Will select first running Pod available
-	it := 0
-	for {
-		if pods.Items[it].Status.Phase == "Running" {
+	var grafanaPod *corev1.Pod
+	for _, pod := range pods.Items {
+		if pod.Status.Phase == "Running" {
+			grafanaPod = &pod
 			break
 		}
-
-		it++
-		if it == len(pods.Items) {
-			return errors.Errorf("No running Grafana pod available.")
-		}
+	}
+	if grafanaPod == nil {
+		return errors.Errorf("No running Grafana pod available")
 	}
 
-	// Build http spdy-upgraded handler
-	dialer := createDialer(conf, v1ClientSet, pods.Items[it].GetName())
-
-	// StopChan is used to understand the lifecycle of the forwarding blocking op
-	// ReadyChan signals when the connection has been established and forwarding is in place
-	stopChan := make(chan struct{}, 1)
-	readyChan := make(chan struct{}, 1)
-	out, errOut := new(bytes.Buffer), new(bytes.Buffer)
-
-	forwardStr := fmt.Sprintf("%d:%d", d.localPort, d.remotePort)
-	fmt.Fprintf(d.out, "[+] Using forwarding: %s\n", forwardStr)
-
-	forwarder, err := portforward.New(dialer,
-		[]string{forwardStr},
-		stopChan,
-		readyChan,
-		out,
-		errOut)
+	portForwarder, err := NewPortForwarder(conf, clientSet, grafanaPod.Name, grafanaPod.Namespace, d.localPort, d.remotePort)
 	if err != nil {
-		return errors.Errorf("Failed to create forwarder: %s\n", err)
+		return errors.Errorf("Error setting up port forwarding: %s", err)
 	}
 
-	// Binding SIGINT & SIGTERM to trigger the closing routine
-	signal.Notify(d.sigintChan, syscall.SIGINT, syscall.SIGTERM)
-	go func() {
-		<-d.sigintChan // Blocking
-		fmt.Fprintf(d.out, "[+] SIGINT/TERM received, closing")
-		close(stopChan)
-	}()
-
-	// This routine blocks on readyChan till forwarding is setup and ready.
-	go func() {
-		<-readyChan
-
-		fmt.Fprintf(d.out, "[+] Port forwarding successful (localhost:%d)\n", d.localPort)
+	err = portForwarder.Start(func(pf *PortForwarder) error {
 		if d.openBrowser {
 			url := fmt.Sprintf("http://localhost:%d", d.localPort)
 			fmt.Fprintf(d.out, "[+] Issuing open browser %s\n", url)
 			browser.OpenURL(url)
 		}
-	}()
-
-	if err = forwarder.ForwardPorts(); err != nil { // Locks until stopChan is closed.
-		return errors.Errorf("Failed to execute Forward: %s\n", err)
+		return nil
+	})
+	if err != nil {
+		return errors.Errorf("Port forwarding failed: %s", err)
 	}
 
+	// The command should only exit when a signal is received from the OS.
+	// Exiting before will result in port forwarding to stop causing the browser
+	// if open to not render the dashboard.
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt)
```

Comment:
Is there a particular reason for not also waiting on `SIGTERM` like before? I don't disagree with the change if that was intentional, just curious.

Reply:
SIGTERM doesn't work on Windows.